tpchgen-cli/src/main.rs

```diff
@@ -61,137 +68,323 @@ enum Table {
+    cli.main()
 }
 
-fn new_table_writer(cli: &Cli, filename: &str) -> io::Result<BufWriter<File>> {
-    let path = cli.output_dir.join(filename);
-    let file = File::create(path)?;
-    Ok(BufWriter::with_capacity(32 * 1024, file))
-}
+impl Cli {
+    fn main(self) -> io::Result<()> {
+        // Create output directory if it doesn't exist
+        fs::create_dir_all(&self.output_dir)?;
+
+        // Determine which tables to generate
+        let tables: Vec<Table> = if let Some(tables) = self.tables.as_ref() {
+            tables.clone()
+        } else {
+            vec![
+                Table::Nation,
+                Table::Region,
+                Table::Part,
+                Table::Supplier,
+                Table::PartSupp,
+                Table::Customer,
+                Table::Orders,
+                Table::LineItem,
+            ]
+        };
 
-fn generate_nation(cli: &Cli) -> io::Result<()> {
-    let filename = "nation.tbl";
-    let mut writer = new_table_writer(cli, filename)?;
+        // Generate each table
+        for table in tables {
+            match table {
+                Table::Nation => self.generate_nation()?,
+                Table::Region => self.generate_region()?,
+                Table::Part => self.generate_part()?,
+                Table::Supplier => self.generate_supplier()?,
+                Table::PartSupp => self.generate_partsupp()?,
+                Table::Customer => self.generate_customer()?,
+                Table::Orders => self.generate_orders()?,
+                Table::LineItem => self.generate_lineitem()?,
+            }
+        }
 
-    let generator = NationGenerator::new();
-    for nation in generator.iter() {
-        writeln!(writer, "{nation}",)?;
+        println!("Generation complete!");
+        Ok(())
     }
-    writer.flush()
-}
 
-fn generate_region(cli: &Cli) -> io::Result<()> {
-    let filename = "region.tbl";
-    let mut writer = new_table_writer(cli, filename)?;
+    /// return the output filename for the given table
+    fn output_filename(&self, table: Table) -> String {
+        let extension = match self.format {
+            OutputFormat::Tbl => "tbl",
+            OutputFormat::Csv => "csv",
+            OutputFormat::Parquet => "parquet",
+        };
+        format!("{}.{extension}", table.name())
+    }
 
-    let generator = RegionGenerator::new();
-    for region in generator.iter() {
-        writeln!(writer, "{region}",)?;
+    /// return a buffered file for writing the given filename in the output directory
+    fn new_output_writer(&self, filename: &str) -> io::Result<BufWriter<File>> {
+        let path = self.output_dir.join(filename);
+        let file = File::create(path)?;
+        Ok(BufWriter::with_capacity(32 * 1024, file))
     }
-    writer.flush()
-}
 
-fn generate_part(cli: &Cli) -> io::Result<()> {
-    let filename = "part.tbl";
-    let mut writer = new_table_writer(cli, filename)?;
+    fn generate_nation(&self) -> io::Result<()> {
```

Comment:
I chose to leave the structure one function per table, which then dispatches to a specialized function for format

We could likely avoid the replication with macros / traits, but I think this way is pretty explicit (and the number of tables will never change)